### PR TITLE
Merge the dead code elimination and dependency graph code for gles and vulkan

### PR DESCRIPTION
**Pull request by Qining**

### CODE REVIEW

**Comment by Qining** [comment]
@ben-clayton @dsrbecky @AWoloszyn 
Dead code elimination transform is moved to: gapis/atom/transform
Dependency graph boilerplate code is moved to: gapis/resolve/dependencygraph
API specific dependency graph code is located in gapis/gfxapi/gles/dependency_graph_behaviour_provider.go and gapis/gfxapi/vulkan/dependency_graph_behaviour_provider.go

The test of dead_code_elimination in gapis/gfxapi/gles/dead_code_elimination_test.go is divided into two parts. First API agnostic liveness tree test, is moved to gapis/atom/transform. Another GLES test stays in gapis/gfxapi/gles/

Two interfaces are added:
1) HasDependencyGraphBehaviourProvider
For APIs that implement this interface, behaviour provider will be requested when building the dependency graph.
If an API does not have dependency graph behaviour provider, all atoms of that API will be labelled: KeepAlive.

2) BehaviourProvider
Provider should: For a given Atom, determine the AtomBehaviour of it and returns the AtomBehaviour. The function: GetBehaviourForAtom() is called when resolving the dependency graph.

Design decision:
Add a new Behaviour Provider interface, because building dependency graph for Vulkan does need some extra book keeping.

Not building the whole dependency graph in the provider, because the atom list may contain atoms from different APIs, so each atom should be processed by its corresponding API. This is why the provider just process one atom, instead of the whole atom list.

Some parts not so good, or can be improved:
GetBehaviourForAtom() does need the dependency graph to be passed in. Ideally, the provider should not need access to the graph at all. This is because the AtomBehaviour describes how 'stateKey' affects the 'stateAddress'. The translation between 'StateKey' and 'StateAddress' is owned by the dependency graph. One way to avoid the graph to be passed into provider is postpone the translation. However I'm not sure if it should be done in this CL.

File: gapis/resolve/dependencygraph/dependency_graph.go [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+	}
+}
+
+func (b *AtomBehaviour) Write(g *DependencyGraph, state StateKey) {
+	if state != nil {
+		b.Writes = append(b.Writes, g.addressMap.addressOf(state))
+	}
+}
+
+type HasDependencyGraphBehaviourProvider interface {
```
Comment (dsrbecky) [inline gapis/resolve/dependencygraph/dependency_graph.go]: I am not sure if the "Has" prefix is very go-like.
Personally, it seems fine to me.  @ben-clayton ?

File: gapis/resolve/dependencygraph/dependency_graph.go [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+	s := c.NewState()
+	t0 := dependencyGraphBuildCounter.Start()
+	for i, a := range g.Atoms {
+		api := a.API()
+		if _, ok := behaviourProviders[api]; !ok {
+			if bp, ok := api.(HasDependencyGraphBehaviourProvider); ok {
+				behaviourProviders[api] = bp.GetDependencyGraphBehaviourProvider(ctx)
+			} else {
+				// API does not provides dependency information, always keep atoms for
```
Comment (dsrbecky) [inline gapis/resolve/dependencygraph/dependency_graph.go]: nit: "provide"

File: gapis/resolve/dependencygraph/dependency_graph.go (hunk 1 above)
Comment (ben-clayton) [inline gapis/resolve/dependencygraph/dependency_graph.go]: No, it's not go like. Please drop the `Has`.

**Comment by Qining** [comment]
Rebased on top of #613

File: gapis/config/config.go [hunk 3]
```diff
@@ -20,7 +20,7 @@ const (
 	DebugReplay                = false
 	DebugReplayBuilder         = false
 	DisableDeadCodeElimination = false
-	DebugDeadCodeElimination   = false
+	DebugDeadCodeElimination   = true
```
Comment (Qining) [inline gapis/config/config.go]: Sorry, this is my fault. Fixed.

File: gapis/resolve/dependencygraph/dependency_graph.go (hunk 1 above)
Comment (Qining) [inline gapis/resolve/dependencygraph/dependency_graph.go]: Dropped in 5aea7b6

File: gapis/resolve/dependencygraph/dependency_graph.go (hunk 2 above)
Comment (Qining) [inline gapis/resolve/dependencygraph/dependency_graph.go]: Fixed in 5aea7b6

**Comment by Qining** [comment]
@dsrbecky @ben-clayton @AWoloszyn 
efde581 is the commit rebased on top of current master with synchronization cl and updated gles dep graph. PTAL.

**Comment by dsrbecky** [comment]
LGTM

**Comment by Qining** [comment]
@dsrbecky 
Replay of GLES fixed at: https://github.com/Qining/gapid/blob/8202971d612951896754485e7a9f03c2b9888eab/gapis/resolve/dependencygraph/dependency_graph.go#L174

The problem is that, GLES atoms may depends on the side effects of Core API atoms. So just keep all the Core API atoms alive is not enough, we needs to mutate them.

**Pull request merged**